File: Hailstone.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class MainClass {
  // Define
  public static IEnumerable<int> Hailstone(int n) 
  {
    if(n == 1) {
      return new [] {1};
    }

    return n % 2 == 0
             ? new [] {n}.Concat(Hailstone(n / 2))
             : new [] {n}.Concat(Hailstone(3 * n + 1));
  }

  // Test
  public static void Main (string[] args) 
  {
    for(var i = 1; i < 100; i++) 
    {
      Console.WriteLine($"{i}: [{string.Join(", ", Hailstone(i))}]");
    }
  }
}

File: hailstone.fs
// Define
let rec hailstone n = 
  match n with 
  | n when n = 1 -> [|1|]
  | n when n % 2 = 0 -> Array.concat [ [|n|] ; hailstone(n / 2) ]
  | n -> Array.concat [ [|n|] ; hailstone(3 * n + 1) ]

// Test
for n in 1..100 do
  hailstone n
  |> Array.map string
  |> String.concat ", "
  |> sprintf "%d: [%s]" n
  |> System.Console.WriteLine

